### PR TITLE
swarm/api/http: http package refactoring

### PR DESCRIPTION
**Pull request by justelad**
this PR addresses a much needed cleanup and refactoring of the swarm `http` package.

the changeset covers:
- appropriate methods have been moved into the `api` package
- introduces middleware-style routing
- serves the `favicon` upon request to `/favicon.ico`
- redesign&cleanup of the html responses (thanks @Mistra !)
- consolidates all responses under `response.go`, templates are in `templates.go`
- compact responses for `curl` requests (!) :)
- introduction of `sctx` as a way to pass information between components using `context` objects

resolves https://github.com/ethersphere/go-ethereum/issues/772
resolves https://github.com/ethersphere/go-ethereum/issues/457
resolves https://github.com/ethersphere/go-ethereum/issues/153

### CODE REVIEW

File: swarm/api/http/response.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/metrics"
+	"github.com/ethereum/go-ethereum/swarm/api"
+)
+
+//metrics variables
```
Comment (nonsense) [inline swarm/api/http/response.go]: Useless comment.

File: swarm/api/http/response.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	})
+}
+
+func RespondError(w http.ResponseWriter, r *http.Request, msg string, code int) {
+	log.Debug("RespondError", "ruid", GetRUID(r.Context()), "uri", GetURI(r.Context()))
+	RespondTemplate(w, r, "error", msg, code)
+}
+
+//evaluate if client accepts html or json response
+func respond(w http.ResponseWriter, r *http.Request, params *ResponseParams) {
```
Comment (nonsense) [inline swarm/api/http/response.go]: Please use godoc.

File: swarm/api/http/response.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+		}
+	} else if strings.Contains(acceptHeader, "text/html") {
+		respondHTML(w, r, params)
+	} else {
+		respondPlaintext(w, r, params) //returns nice errors for curl
+	}
+}
+
+//return a HTML page
+func respondHTML(w http.ResponseWriter, r *http.Request, params *ResponseParams) {
```
Comment (nonsense) [inline swarm/api/http/response.go]: Please use godoc.

File: swarm/api/http/templates.go [hunk 4]
```diff
@@ -28,178 +29,269 @@ type htmlListData struct {
...
+  color: #222;
+  padding: 15px 32px;
+  text-align: center;
+  text-decoration: none;
+  display: inline-block;
+  font-size: 16px;
+}
+{{ end }}`
+
+const logo = `{{ define "logo" }}
```
[the remark this answers is not in the document]
Comment (justelad) [reply swarm/api/http/templates.go]: Have a look at the go templating engine documentation. There are also pretty good tutorials out there. I didn’t do anything fancy, just iterated again and again to minimize the footprint of the code and the templates

**Pull request merged**